src/runtime/sources.cpp

```diff
@@ -254,131 +254,66 @@ static std::string slurp(int dirfd, const char *const *argv, bool &fail) {
-      std::string submodStr;
-      if (faccessat(dirfd, ".gitmodules", R_OK, 0) == 0) {
-        static const char *submodArgs[] = {
-            "git",  "config", "-f", ".gitmodules", "-z", "--get-regexp", "^submodule[.].*[.]path$",
-            nullptr};
-        submodStr = slurp(dirfd, &submodArgs[0], failed);
-      }
-      std::string prefix(path == "." ? "" : (path + "/"));
-      const char *tok = fileStr.data();
-      const char *end = tok + fileStr.size();
-      for (const char *scan = tok; scan != end; ++scan) {
-        if (*scan == 0 && scan != tok) {
-          files.emplace_back(prefix + tok);
-          tok = scan + 1;
-        }
-      }
-      tok = submodStr.data();
-      end = tok + submodStr.size();
-      const char *value = nullptr;
-      for (const char *scan = tok; scan != end; ++scan) {
-        if (*scan == '\n' && !value) value = scan + 1;
-        if (*scan == 0 && scan != tok && value) {
-          submods.emplace_back(prefix + value);
-          tok = scan + 1;
-          value = nullptr;
-        }
-      }
-    } else {
-      bool recurse;
-#ifdef DT_DIR
-      if (f->d_type == DT_UNKNOWN) {
-#endif
-        struct stat sbuf;
-        if (fstatat(dirfd, f->d_name, &sbuf, AT_SYMLINK_NOFOLLOW) != 0) {
-          fprintf(stderr, "Failed to fstatat %s/%s: %s\n", path.c_str(), f->d_name,
-                  strerror(errno));
-          failed = true;
-          recurse = false;
-        } else {
-          recurse = S_ISDIR(sbuf.st_mode);
-        }
-#ifdef DT_DIR
-      } else {
-        recurse = f->d_type == DT_DIR;
-      }
-#endif
-      if (recurse) {
-        std::string name(path == "." ? f->d_name : (path + "/" + f->d_name));
-        if (name == ".build" || name == ".fuse") continue;
-        int fd = openat(dirfd, f->d_name, O_RDONLY);
-        if (fd == -1) {
-          fprintf(stderr, "Failed to openat %s/%s: %s\n", path.c_str(), f->d_name, strerror(errno));
-          failed = true;
-        } else {
-          if (scan(files, submods, name, fd)) failed = true;
-        }
-      }
-    }
-  }
+  return S_ISDIR(sbuf.st_mode);
+}
+
+std::vector<std::string> scan_git(int dirfd) {
+  std::vector<std::string> files;
+
+  bool failed;
+  static const char *fileArgs[] = {"git", "ls-files", "-z", "--recurse-submodules", nullptr};
+  std::string fileStr(slurp(dirfd, &fileArgs[0], failed));
 
-  if (errno != 0 && !failed) {
-    fprintf(stderr, "Failed to readdir %s: %s\n", path.c_str(), strerror(errno));
-    failed = true;
+  if (failed) {
+    // TODO: log a nice message here
+    exit(1);
   }
 
-  if (closedir(dir) != 0) {
-    fprintf(stderr, "Failed to closedir %s: %s\n", path.c_str(), strerror(errno));
-    failed = true;
+  const char *tok = fileStr.data();
+  const char *end = tok + fileStr.size();
+
+  for (const char *scan = tok; scan != end; ++scan) {
+    if (*scan == 0 && scan != tok) {
+      std::string path(tok);
+      tok = scan + 1;
+
+      if (!is_directory(path)) {
```

Comment:
We decided not to filter this out didn't we? Also this means that you stat every source up front when you could just wait for the inode to be touched when a user calls `source` or `sources` instead. The only downside is that empty directories from inactive sub-modules will be in `sources` but that seems fine to me.

Reply:
oh yeah,  sorry. Need to remove this